packages/admin-ui/src/text-area/text-area.tsx

```diff
@@ -0,0 +1,136 @@
+import type { ReactNode, ComponentPropsWithoutRef, Ref } from 'react'
+import React, { useRef, useState, forwardRef } from 'react'
+import { useId } from '@vtex/admin-ui-hooks'
+
+import { TextAreaContainer } from './text-area-container'
+import { TextAreaElement } from './text-area-element'
+import { Text } from '../components/Text'
+
+import {
+  FormControl,
+  FormControlLabel,
+  FormControlMessage,
+} from '../form-control'
+import { Flex, FlexSpacer } from '..'
+import { MAX_HEIGHT } from './text-area.style'
+
+export const TextArea = forwardRef(
+  (props: TextAreaProps, ref: Ref<HTMLTextAreaElement>) => {
+    const {
+      error,
+      disabled,
+      errorText,
+      helpText,
+      label,
+      charLimit,
+      id: defaultId,
+      onChange,
+      ...inputProps
+    } = props
+
+    const id = useId(defaultId)
+    const containerRef = useRef<any>(null)
+
+    const [charCount, setCharCount] = useState(0)
+    const [lineHeight, setLineHeight] = useState<number>(0)
+    const [minScrollHeight, setMinScrollHeight] = useState<number>(0)
+
+    const getLineHeight = (elm: HTMLTextAreaElement) => {
+      const savedRows = elm.rows
+      const savedValue = elm.value
+
+      elm.value = ''
+      elm.rows = 1
+      const base1 = elm.scrollHeight
+
+      elm.rows = 2
+      const base2 = elm.scrollHeight
+
+      elm.rows = savedRows
+      elm.value = savedValue
+
+      const calculatedHeight = base2 - base1
+
+      setLineHeight(calculatedHeight)
+
+      return calculatedHeight
+    }
+
+    const getScrollHeight = (elm: HTMLTextAreaElement) => {
+      const savedValue = elm.value
+
+      elm.value = ''
+
+      const baseScrollHeight = elm.scrollHeight
+
+      elm.value = savedValue
+
+      setMinScrollHeight(baseScrollHeight)
+
+      return baseScrollHeight
+    }
+
+    const onExpandableTextareaInput = (
+      e: React.FormEvent<HTMLTextAreaElement>
+    ) => {
+      const elm = e.target as HTMLTextAreaElement
+      const baseRows = 2
+
+      lineHeight || getLineHeight(elm)
+      minScrollHeight || getScrollHeight(elm)
```

Comment:
These lines are just confusing to me. What are you trying to do here?

Reply:
Yeah, it's a little weird, i think i'll change it to be a little more "reactic".  It was based on that codepen you sent me and it didn't fit well

Here if this state wasn't set it would trigger a setstate the first time, and then not do the operation again